animation.cpp

```diff
@@ -72,3 +72,78 @@ PAnimation Animation::fromId(AnimationId id) {
         {{577, 598}, {94, 94}, {-29, -29}},
         }));
 }
+
+class Particle {
+  public:
+  Particle(Vec2 position, Vec2 origin, Color color, Vec2 speed, int size)
+      : position(position), origin(origin), color(color), speed(speed), size(size) {
+    color.a = 128;
+  }
+
+  void update() {
+    // custom particle effect logic
+    position +=speed;
+
+    if (size > 1) {
+      size--;
+    }
+
+    color.r -= 10;
+    color.g -= 10;
+    color.b -= 10;
+    color.a -= 10;
+  }
+
+  void draw(Renderer& renderer) {
+    renderer.drawPoint(position + origin, color, size);
+  }
+
+  private:
+  Vec2 position;
+  Vec2 origin;
+  Color color;
+  Vec2 speed;
+  int size;
+
+  friend class ParticleEffect;
+  };
+
+class ParticleEffect : public Animation {
+  public:
+  ParticleEffect(milliseconds duration, unsigned int particleNum, Vec2 origin)
```

Comment:
It seems that the particleNum parameter is unnecessary. It's overwritten to 8 just below, and actually never read.